### PR TITLE
Refactor/extract

### PR DESCRIPTION
**Pull request by vgainullin**
This pull request will refactor extract module into Extract and Data classes.

### CODE REVIEW

File: txt2hpo/extract.py [hunk 1]
```diff
@@ -11,7 +11,232 @@
...
 
-context_model = load_model()
+
+class Data(object):
+    def __init__(self, entries=None, model=None):
+        if not entries:
+            self.entries = []
+        else:
+            self.entries = entries
+        self.model = model
```
Comment (arvkevi) [inline txt2hpo/extract.py]: validate `model` with a conditional or `try`, `except`.

File: txt2hpo/extract.py [hunk 2]
```diff
@@ -11,7 +11,232 @@
...
+
+            resolved_terms.append(entry)
+        self.entries = resolved_terms
+
+    @property
+    def hpids(self):
+        return [x['hpid'] for x in self.entries]
+
+    @property
+    def json(self):
```
Comment (arvkevi) [inline txt2hpo/extract.py]: I'd suggest renaming this function, `json` is standard library and you might run into namespace collisions.

File: txt2hpo/extract.py [hunk 3]
```diff
@@ -320,12 +352,12 @@ def self_evaluation(correct_spelling=False, resolve_conflicts=False):
     logger.info('Running self evaluation, this may take a few minutes \n')
     i = 0
     n_nodes = len(hpo_network.nodes)
-
+    hpo = Extractor(correct_spelling=correct_spelling,resolve_conflicts=resolve_conflicts).hpo
```
Comment (arvkevi) [inline txt2hpo/extract.py]: `extractor = Extractor(correct_spelling=correct_spelling,resolve_conflicts=resolve_conflicts)`

File: txt2hpo/extract.py [hunk 4]
```diff
@@ -320,12 +352,12 @@ def self_evaluation(correct_spelling=False, resolve_conflicts=False):
...
     for node in hpo_network:
         total += 1
         term = hpo_network.nodes[node]['name']
         hpids = []
-        extracted = hpo(term, correct_spelling=correct_spelling, resolve_conflicts=resolve_conflicts)
+        extracted = hpo(term)
```
Comment (arvkevi) [inline txt2hpo/extract.py]: `extractor.hpo(term)`

File: tests/test_extract.py [hunk 5]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
 
     def test_hpo(self):
 
-        # Test extracting an abbreviated phenotype
-        truth = json.dumps([{"hpid": ["HP:0001370"], "index": [0, 2], "matched": "RA"}])
-        self.assertEqual(hpo("RA"), truth)
+        hpo = Extractor(correct_spelling=False).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: suggesting changing this to `extractor = Extractor(correct_spelling=False)`. Then use `extractor.hpo(term)` later

File: tests/test_extract.py [hunk 6]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
...
-                             "context": "female with fourth metacarpal brachydactyly"}])
+        truth = [{"hpid": ["HP:0001156"], "index": [30, 43], "matched": "brachydactyly",
+                             "context": "Female with fourth metacarpal brachydactyly"}]
 
-        self.assertEqual(truth, hpo("Female with fourth metacarpal brachydactyly", return_context=True))
+        self.assertEqual(truth, hpo("Female with fourth metacarpal brachydactyly").entries)
 
-        truth = json.dumps([{"hpid": ["HP:0000988"], "index": [23, 27], "matched": "rash"},
-                            {"hpid": ["HP:0000988"],"index": [18, 27], "matched": "skin rash"},
+        hpo = Extractor(correct_spelling=False).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 7]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
...
-                             "context": "developmental and delay"}])
-        self.assertEqual(hpo("developmental and delay", max_neighbors=3, return_context=True), truth)
-        truth = json.dumps([])
-        self.assertEqual(hpo("developmental and delay", max_neighbors=1, return_context=True), truth)
+        hpo = Extractor(correct_spelling=True, max_neighbors=3).hpo
+        truth = [{"hpid": ["HP:0001263"], "index": [0, 23], "matched": "developmental and delay",
+                             "context": "developmental and delay"}]
+        self.assertEqual(hpo("developmental and delay").entries, truth)
+
+        hpo = Extractor(correct_spelling=True, max_neighbors=1).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 8]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
...
+        truth = []
+        self.assertEqual(hpo("developmental and delay").entries, truth)
+
+        hpo = Extractor(correct_spelling=False).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 9]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
...
-                            ])
-        self.assertEqual(hpo("developmental delay, hypotonia", max_length=20), truth)
+        # Test extracting multiple phenotypes
+        truth = [{"hpid": ["HP:0001290"], "index": [0, 9], "matched": "Hypotonia"},
+                            {"hpid": ["HP:0001263"], "index": [11, 30], "matched": "developmental delay"
+                             }]
+        self.assertEqual(hpo("Hypotonia, developmental delay").entries_sans_context, truth)
+
+        # Test term indexing given max length of extracted text
+        hpo = Extractor(correct_spelling=False, max_length=20).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 10]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
...
     def test_hpo_big_text_spellcheck_on(self):
         # test parsing a page
-        self.assertEqual(len(json.loads(hpo(test_case11_text, max_neighbors=2))), 9)
+        hpo = Extractor(max_neighbors=2).hpo
+        self.assertEqual(hpo(test_case11_text).n_entries, 10)
 
     def test_hpo_big_text_spellcheck_off(self):
         # test parsing a page
-        self.assertEqual(len(json.loads(hpo(test_case11_text, max_neighbors=2, correct_spelling=False))), 9)
+        hpo = Extractor(max_neighbors=2, correct_spelling=False).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 11]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
...
         # test parsing a page
-        self.assertEqual(len(json.loads(hpo(test_case11_text, max_neighbors=3, correct_spelling=False))), 10)
+        hpo = Extractor(max_neighbors=3, correct_spelling=False).hpo
+        self.assertEqual(hpo(test_case11_text).n_entries, 11)
 
     def test_hpo_big_text_max_neighbors(self):
         # test parsing a page
-        hpo_max_2 = json.loads(hpo(test_case11_text, max_neighbors=2))
-        hpo_max_3 = json.loads(hpo(test_case11_text, max_neighbors=3))
+        hpo = Extractor(max_neighbors=2, correct_spelling=True).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 12]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
...
+        hpo_max_2 = hpo(test_case11_text).hpids
+        hpo = Extractor(max_neighbors=3, correct_spelling=True).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 13]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
...
+        hpo_max_3 = hpo(test_case11_text).hpids
 
         self.assertNotEqual(hpo_max_2, hpo_max_3)
 
     def test_iteration_over_chunks(self):
         # test performing multiple extractions in a row
         sentences = ['Developmental delay', 'Hypotonia']
+        hpo = Extractor(correct_spelling=False).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 14]
```diff
@@ -21,192 +21,199 @@ def test_group_sequence(self):
...
-            self.assertNotEqual(len(result), 0)
+            self.assertNotEqual(hpo(sentence).n_entries, 0)
         sentences = ['Hypotonia', 'Developmental delay']
-
         for sentence in sentences:
-            result = json.loads(hpo(sentence, correct_spelling=False))
-            self.assertNotEqual(len(result), 0)
-
+            self.assertNotEqual(hpo(sentence).n_entries, 0)
+        hpo = Extractor(correct_spelling=True).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 15]
```diff
@@ -271,4 +284,47 @@ def test_conflict_resolver(self):
...
-        self.assertEqual(truth, conflict_resolver(extracted, context_model))
+        data = Data(entries=extracted, model=model)
+        data.resolve_conflicts()
+        self.assertEqual(truth, data.entries)
+
+    def test_custom_synonyms(self):
+        # test adding custom synonyms
+        custom_syn = {"HP:0001263": ['DD', 'GDD'], "HP:0000729": ['ASD', 'PDD']}
+
+        hpo = Extractor(custom_synonyms=custom_syn).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 16]
```diff
@@ -271,4 +284,47 @@ def test_conflict_resolver(self):
...
+        truth = [{"hpid": ["HP:0001263"], "index": [0, 3], "matched": "GDD"},
+                            {"hpid": ["HP:0001263"], "index": [4, 6], "matched": "DD"}]
+        self.assertEqual(hpo("GDD DD").entries_sans_context, truth)
+
+    def test_extract_ambiguous(self):
+        # test resolver works
+        hpo = Extractor(resolve_conflicts=True).hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: tests/test_extract.py [hunk 17]
```diff
@@ -271,4 +284,47 @@ def test_conflict_resolver(self):
...
+        res = ex.hpo("ASD")
+        self.assertEqual(len(res.entries[0]['hpid']), 2)
+
+        ex = Extractor(resolve_conflicts=True)
+        self.assertEqual(ex.resolve_conflicts, True)
+        res = ex.hpo("ASD")
+        self.assertEqual(len(res.entries[0]['hpid']), 1)
+
+    def test_extract_from_repeated_context(self):
+        hpo = Extractor().hpo
```
Comment (arvkevi) [inline tests/test_extract.py]: see comment above

File: txt2hpo/extract.py (hunk 2 above)
Comment (vgainullin) [reply txt2hpo/extract.py]: requests has a json method, its very useful for them.

Comment (vgainullin) [reply txt2hpo/extract.py]: response = requests.get()
response.json
response.text 
is the pattern i was going for

File: txt2hpo/extract.py (hunk 4 above)
Comment (vgainullin) [reply txt2hpo/extract.py]: yeah, that seems good, I just wanted to fit those into existing tests as much as possible

**Pull request merged**